L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+  caloParticlesToken_(consumes<CaloParticleCollection>(_config.getParameter<edm::InputTag>("caloParticles"))),
+  triggerCellsToken_(consumes<l1t::HGCalTriggerCellBxCollection>(_config.getParameter<edm::InputTag>("triggerCells"))),
+  // simHitsTokenEE_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsEE"))),
+  // simHitsTokenHEfront_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEfront"))),
+  // simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEback")))
+  dummyClustering_(_config.getParameterSet("dummyClustering"))
+{
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>("HGCalBackendLayer1Processor2DClustering");
+  produces<l1t::HGCalMulticlusterBxCollection>("HGCalBackendLayer2Processor3DClustering");
+  if (makeCellsCollection_)
+    produces<l1t::HGCalTriggerCellBxCollection>("HGCalVFEProcessorSums");
+}
+
+CaloTruthCellsProducer::~CaloTruthCellsProducer()
+{
+}
+
+void
+CaloTruthCellsProducer::produce(edm::Event& _event, edm::EventSetup const& _setup)
+{
+  edm::Handle<CaloParticleCollection> caloParticlesHandle;
+  _event.getByToken(caloParticlesToken_, caloParticlesHandle);
+  auto& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  _event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto& triggerCells(*triggerCellsHandle);
+
+  // std::map<uint32_t, double> hitMap(makeHitMap(_event, _setup)); // cellId -> sim energy
+  // std::map<uint32_t, double> tcMap; // tcId -> total sim energy
+
+  dummyClustering_.eventSetup(_setup);
+  showerShape_.eventSetup(_setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  _setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto& geometry(*geometryHandle);
+
+  std::map<uint32_t, CaloParticleRef> tcToCalo;
+
+  // used later to order multiclusters
+  std::vector<CaloParticleRef> orderedCaloRefs;
+
+  for (unsigned iP(0); iP != caloParticles.size(); ++iP) {
+    auto& caloParticle(caloParticles.at(iP));
+    if (caloParticle.g4Tracks().at(0).eventId().event() != 0) // pileup
+      continue;
+
+    CaloParticleRef ref(caloParticlesHandle, iP);
+
+    SimClusterRefVector const& simClusters(caloParticle.simClusters());
+    for (auto& simCluster : simClusters) {
+      for (auto& hAndF : simCluster->hits_and_fractions()) {
+        DetId hitId(hAndF.first);
+        HGCalDetId tcId;
+        try {
+          tcId = HGCalDetId(geometry.getTriggerCellFromCell(hitId));
+        }
+        catch (cms::Exception const& ex) {
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        // tcMap[tcId] += hitMap[hitId.rawId()] * hAndF.first;
+        tcToCalo.emplace(tcId, ref);
+      }
+    }
+
+    // ordered by the gen particle index
+    int genIndex(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
```

Comment:
Having a `set` of `CaloParticleRef` would simplify the code here and below

Reply:
Please see above. The index in the CaloParticleCollection does not correspond to the genParticle index the calo particle has.